TECHMANIA/Assets/Scripts/Components/Game Scene/Game.cs

```diff
@@ -1717,10 +1717,11 @@ private void JumpToScan(int scan)
                 AudioClip clip = ResourceLoader.GetCachedClip(
                     n.sound);
                 if (clip == null) return;
-                if (n.time + clip.length > BaseTime)
+                if (n.time < BaseTime &&
```

Comment:
I see the issue that keysounds may be played twice if `n.time == BaseTime`, but it's more nuanced than that.

`Game.UpdateTime` only plays unremoved notes in `noteObjectsInLane`, and this method removes notes in a pretty peculiar way (see `NoteList.RemoveUpTo`). A more proper and easier fix would be for this `foreach` loop to only consider removed notes in `noteObjectsInLane`, so that no note will be played by both this loop and `Game.UpdateTime`. To do that:

* Add a method to `NoteList` that only runs on removed items, for example `ForEachRemoved`
* In this `foreach` loop, call `l.ForEachRemoved` instead of `l.ForEach`, and in the action there's no need for `if (n.time > BaseTime) return;`
* The off-by-1 fix should of course remain

Reply:
Your suggested approach requires a little change in `NoteList.RemoveUpTo` regarding end of scan notes since there are hidden notes generated by auto keysound as well. With that, I included lane checking into the code block.

Reply:
For auto-keysound notes, whether they are end-of-scan doesn't matter, as long as exactly one of `Game.UpdateTime` and `Game.JumpToScan` plays them. Therefore the additional check in `NoteList.RemoveUpTo` shouldn't be necessary. But please give it a test just in case.

Reply:
About that, I have tested. The test cases and results on jumping to end of scan notes (hidden or playable) are as follow.

- With `NoteList.RemoveUpTo` untouched, both hidden and playable notes are ignored. The playable one is not included as a note to hit, but the hidden one does not play its keysound.
- With End of Scan check in `NoteList.RemoveUpTo` removed altogether, the hidden one plays, but the playable one still gets resolved.
- With the current change on this branch, it works as expected.

The same set of behaviors described above also apply to normal keysound as well. It worked in the current official build because `Game.UpdateTime` plays the keysound while `Game.JumpToScan` does not. Notice how it plays the sound once, and not twice as I described the original issue in the subject.

Reply:
Just so we are talking about the same change, I'm referring to your change from `beforeThreshold = list[i].note.endOfScan;` to `beforeThreshold = list[i].note.endOfScan && list[i].note.lane < Pattern.kAutoKeysoundFirstLane;`, and I argue that the `list[i].note.lane < Pattern.kAutoKeysoundFirstLane` condition you added is unnecessary; `list[i].note.endOfScan` should remain.

I tested on my end and no sound plays twice, whether playable or not, whether end-of-scan or not, whether auto-keysound or not.

Reply:
Yes, we were talking about the same change, but not the same commit chain. I have traced it and tested; They all play once. However, once I applied `&& n.note.lane < Pattern.kAutoKeysoundFirstLane;` into `PlayKeysound` ([Game.cs:2439](https://github.com/techmania-team/techmania/pull/18/commits/63c42b91b40f7b56d07a4342a2748932cf1417c7)) which handles auto keysound volumes, I went as I described.

Reply:
That led me to find another issue: we fixed how we calculate `hidden` in `Game.PlayKeysound`, but not in `Game.JumpToScan` when it calls `audioSourceManager.PlayKeysound`. We need to add a `&& n.note.lane < Pattern.kAutoKeysoundFirstLane` here as well.

Anyway, on master, plus `l.ForEach` to `l.ForEachRemoved`, plus the 2 `&& n.note.lane < Pattern.kAutoKeysoundFirstLane`s, I still can't replicate your result, the "the hidden one does not play its keysound" part. When jumping to an end-of-scan note, `RemoveUpTo` removes it from `noteObjectsInLane`, then `Game.JumpToScan` plays its keysound, at the correct volume, whether it's a playable note, hidden note or auto keysound note.

Reply:
Alright, this is a recorded case of what happened on my end. This time, it is on master with...

- `ForEachRemoved`
- `&& n.note.lane < Pattern.kAutoKeysoundFirstLane` added in `Game.PlayKeysound`
- `&& n.lane < Pattern.kAutoKeysoundFirstLane` added in `Game.JumpToScan`, next to `if (n.time + clip.length > BaseTime`
- No change on `NoteList.RemoveUpTo`

https://www.youtube.com/watch?v=HEuwi3xFTnE

Should the game play the keysound of the end of scan note jumped onto in this case? The behavior in the video is what I was talking about since my first reply here, but I suppose this does not occur on your end, correct? If that is the case, would you like both `&& n[.note].lane < Pattern.kAutoKeysoundFirstLane`s checked or just removed them?

Reply:
It should and does play on my end, and I honestly don't know what's going on anymore. We have spent too much time on this already; I'll just merge as is. Thank you for your patience.

Reply:
No problem, and thanks for your patience too. If it works on your end, it is all that matters since you are the one who is going to create official builds in the end. We can work on this later if you want.